src/functional/ffront/symbol_makers.py

```diff
@@ -0,0 +1,192 @@
+            if isinstance(dim_arg, list):
+                for d in dim_arg:
+                    if not isinstance(d, common.Dimension):
+                        raise FieldOperatorTypeError(f"Invalid field dimension definition '{d}'")
+                    dims.append(foast.Dimension(name=d.value))
+            elif dim_arg is Ellipsis:
+                dims = dim_arg
+            else:
+                raise FieldOperatorTypeError(f"Invalid field dimensions '{dim_arg}'")
+
+            try:
+                dtype = recursive_make_symbol(dtype_arg)
+            except FieldOperatorTypeError as error:
+                raise FieldOperatorTypeError(
+                    "Field dtype argument must be a scalar type (got '{dtype}')!"
+                ) from error
+            if not isinstance(dtype, foast.ScalarType):
+                raise FieldOperatorTypeError(
+                    "Field dtype argument must be a scalar type (got '{dtype}')!"
+                )
+            return foast.FieldType(dims=dims, dtype=dtype)
+
+        case collections.abc.Callable:
+            if not args:
+                raise FieldOperatorTypeError("Not annotated functions are not supported!")
+
+            try:
+                arg_types, return_type = args
+                args = [recursive_make_symbol(arg) for arg in arg_types]
+            except Exception as error:
+                raise FieldOperatorTypeError(
+                    f"Invalid callable annotations in {type_hint}"
+                ) from error
+
+            kwargs_info = [arg for arg in extra_args if isinstance(arg, typingx.CallableKwargsInfo)]
+            if len(kwargs_info) != 1:
+                raise FieldOperatorTypeError(f"Invalid callable annotations in {type_hint}")
+            kwargs = {
+                arg: recursive_make_symbol(arg_type)
+                for arg, arg_type in kwargs_info[0].data.items()
+            }
+
+            return foast.FunctionType(
+                args=args, kwargs=kwargs, returns=recursive_make_symbol(return_type)
+            )
+
+    raise FieldOperatorTypeError(f"'{type_hint}' type is not supported")
+
+
+def make_symbol_from_value(
+    name: str, value: Any, namespace: foast.Namespace, location: SourceLocation
+) -> foast.Symbol:
+    if not isinstance(value, type) or type(value).__module__ != "typing":
+        value = typingx.get_typing(value, annotate_callable_kwargs=True)
+    symbol_type = make_symbol_type_from_typing(value)
+    if isinstance(symbol_type, foast.DataType):
+        return foast.DataSymbol(id=name, type=symbol_type, namespace=namespace, location=location)
+    elif isinstance(symbol_type, foast.FunctionType):
+        return foast.Function(
+            id=name,
+            type=symbol_type,
+            namespace=namespace,
+            location=location,
+        )
+    else:
+        raise common.GTTypeError(f"Impossible to map '{value}' value to a Symbol")
+
+
+class FieldOperatorTypeError(common.GTTypeError):
```

Comment:
Will it be possible to recover the ability to tell the user where (source location) the type error comes from?

Reply:
It should be possible and we should do it. I added a TODO to do it in a following PR.